helix-core/src/main/java/org/apache/helix/messaging/handling/HelixTaskExecutor.java

```diff
@@ -1135,6 +1145,34 @@ private void removeMessageFromZK(HelixDataAccessor accessor, Message message,
     }
   }
 
+  private void startMessageRefreshThread() {
+    int periodicMessageRefreshInterval =
+        HelixUtil.getSystemPropertyAsInt(SystemPropertyKeys.PERIODIC_MESSAGE_REFRESH_INTERVAL, -1);
+    if (periodicMessageRefreshInterval > 0) {
+      _messageRefreshTimer = new Timer();
+      try {
+        {
+          if (_manager == null) {
+            throw new HelixException("Periodic message refresh skipped. No helix manager is set.");
+          }
+        }
+        TimerTask messageRefreshTask = new TimerTask() {
+          @Override
+          public void run() {
+            // Trigger read messages from zk
+            NotificationContext changeContext = new NotificationContext(_manager);
+            changeContext.setChangeType(HelixConstants.ChangeType.MESSAGE);
+            onMessage(_manager.getInstanceName(), null, changeContext);
+          }
+        };
+        _messageRefreshTimer.schedule(messageRefreshTask, periodicMessageRefreshInterval,
+            periodicMessageRefreshInterval);
```

Comment:
Good suggestion. In current design, I put the check in invoke method, and if the onMessage (in the case of messages) is not called for a period of time, it will do the refresh.